### PR TITLE
vcpkg manifest mode

### PR DESCRIPTION
**Pull request by Be-ing**
This was really useful to debug #4225 locally. We should not merge this until we get vcpkg binary caching working because it interferes with using the `vcpkg archive` ZIP archives.

### CODE REVIEW

**Comment by poelzi** [comment]
Since we apperently decided to remove packaging files as much as possible, see endless discussion on #3258 , there is no other choice then say very loud NACK from me.

**Comment by uklotzde** [comment]
> Since we apperently decided to remove packaging files as much as possible, see endless discussion on #3258 , there is no other choice then say very loud NACK from me.

This is a different use case as vcpkg is now mandatory for building on Windows and macOS. NixOS is not part of the cross-plattform build toolchain.

**Comment by Be-ing** [comment]
Having this file in the repository is required to use vcpkg's binary caching feature which will allow macOS and Windows developers to stop downloading a > 1 GB ZIP file containing every single dependency every time an update is made to vcpkg. It will also allow us to have Qt5 and Qt6 builds on CI on the same branch and simplify (or maybe allow us to remove) the platform-specific buildenv scripts.

**Comment by poelzi** [comment]
@Be-ing You know, I'm all for developer suppert/enduser support. I'm all for making it as easy as possible, I just don't like hypocracy, that's why I NACK....

**Comment by Swiftb0y** [comment]
@poelzi please elaborate, intervening without even explaining why is not constructive.

**Comment by Be-ing** [comment]
I don't think this is hypocritical. vcpkg is not the same as Nix or other Linux package managers. vcpkg is not designed for distributing end user applications; it is only for C & C++ libraries and development tools like Qt's `moc` and `uic` tools. Nix has infrastructure in place for unofficial user packages for distributing applications so I think that's where Nix packaging files should go. I don't see anyone asking to add the Arch User PKGBUILDs to Mixxx or other applications' upstream repositories; I don't know why Nix should be different.

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "vcpkg"]
+	path = vcpkg
+	url = https://github.com/mixxxdj/vcpkg.git
```
Comment (daschuer) [inline .gitmodules:3]: I don't think using vcpkg as submodule works well. I have bad experiences using submodules in general.
In this case this will clutter the source tree with a huge number of files, only some users really need. This also interferrs when using git worktree.
Please let's stick to our scripted solution in the tool files and a vcpkg tree outside the mixxx source tree.

**Comment by daschuer** [comment]
When I think to casual contributors, I can imagine that it is still the fastest and safest way to download the pre-build environment and use it. 
Many of us have a fast internet connection anyway, at least faster than building the vcpkg environment at home which also means to download many referenced source code repositories and having the space of the temporary build results. 

It is fine, if we add this to our GitHub CI, if it is not too fragile, but lest stick to the binary environment download for local builds by default.

File: .gitmodules (hunk 1 above)
Comment (Be-ing) [reply .gitmodules:3]: If you don't need the submodule then simply don't clone it. It is only needed for macOS and Windows.

Our current solution is not optimal to maintain. Often when starting a new build, the cache has expired since the last build. If there are any build failures, it may waste hours and the packages that did successfully build need to be built all over again. Also there are issues with packages not always getting rebuilt after making a change. In my experience that is not an issue in manifest mode because vcpkg hashes the contents of the port for a cache key.

Using manifest mode will also allow us to easily maintain Qt5 and Qt6 builds simultaneously on one branch.

**Comment by Be-ing** [comment]
My intention here is for the dependencies built on GitHub Actions to be available for local builds through GitHub Packages. However that is not working yet because of a permissions issue. Please double check that GITHUB_TOKEN has write permission to packages.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -125,36 +129,20 @@ jobs:
...
-    - name: "[macOS] Set up cmake"
-      uses: jwlawson/actions-setup-cmake@v1.9
-      # Ubuntu 20.04 should use the CMake version from the repos, and Visual
-      # Studio on Windows comes with its own CMake version anyway.
-      if: runner.os == 'macOS'
-      with:
-        # This should always match the minimum required version in
-        # our CMakeLists.txt
-        cmake-version: '3.16.x'
-
```
Comment (daschuer) [inline .github/workflows/build.yml:137]: This change is unrelated to this PR.
If we remove it we have no check if our minimum CMake dependent still applies.
If you like to remove it, let's do it separately.

Comment (Be-ing) [reply .github/workflows/build.yml:137]: It is already not used for Windows nor Ubuntu. If we do decide to keep it for some OS, macOS is not a good choice because macOS 11 requires a newer version of CMake.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -192,11 +180,32 @@ jobs:
...
-    - name: "Set up build environment"
-      id: buildenv
-      run: ${{ matrix.buildenv_script }} setup
-      env:
-        BUILDENV_BASEPATH: ${{ matrix.buildenv_basepath }}
+    - name: Install Nuget
+      uses: nuget/setup-nuget@v1
+      if: runner.os != 'Linux'
+      with:
+        nuget-version: 5.10.0
```
Comment (daschuer) [inline .github/workflows/build.yml:187]: Nuget 5.11.0.10 is already installed on GitHub's Windows Server 2019 and Nuget 5.6.0.6489 on macOS 10.15

File: .gitmodules (hunk 1 above)
Comment (Be-ing) [reply .gitmodules:3]: It is also easier to test changes to vcpkg packages with local Mixxx builds in manifest mode.

Comment (Be-ing) [reply .gitmodules:3]: Using manifest mode with the submodule also ensures that Windows and macOS developers are using updated dependencies. Currently the buildenv script has to be manually rerun whenever the dependencies are updated.

**Comment by Be-ing** [comment]
Windows build fails with protobuf??

```
[198/201] cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_dll --intdir=CMakeFiles\libprotobuf.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~2\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe  @CMakeFiles\libprotobuf.rsp  /out:libprotobufd.dll /implib:libprotobufd.lib /pdb:libprotobufd.pdb /dll /version:3.15 /machine:x64 /nologo    /debug /INCREMENTAL  && cd ."
FAILED: libprotobufd.dll libprotobufd.lib 
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_dll --intdir=CMakeFiles\libprotobuf.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~2\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe  @CMakeFiles\libprotobuf.rsp  /out:libprotobufd.dll /implib:libprotobufd.lib /pdb:libprotobufd.pdb /dll /version:3.15 /machine:x64 /nologo    /debug /INCREMENTAL  && cd ."
LINK Pass 1: command "C:\PROGRA~2\MICROS~2\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe @CMakeFiles\libprotobuf.rsp /out:libprotobufd.dll /implib:libprotobufd.lib /pdb:libprotobufd.pdb /dll /version:3.15 /machine:x64 /nologo /debug /INCREMENTAL /MANIFEST /MANIFESTFILE:CMakeFiles\libprotobuf.dir/intermediate.manifest CMakeFiles\libprotobuf.dir/manifest.res" failed (exit code 0) with the following output:
Access violation
ninja: build stopped: subcommand failed.
```
I don't know what to make of that error... is that a bug in the linker causing a segfault??

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -192,11 +180,32 @@ jobs:
...
+
+    - name: "Authenticate to GitHub Packages"
+      if: runner.os != 'Linux' && github.repository_owner == 'mixxxdj'
+      shell: bash
+      run: |
+        nuget sources add -Name mixxx -Source https://nuget.pkg.github.com/mixxxdj/index.json -Username mixxxdj -Password ${{ secrets.GITHUB_TOKEN }} -StorePasswordInClearText -ForceEnglishOutput -NonInteractive
+        nuget setapikey ${{ secrets.GITHUB_TOKEN }} -Source mixxx -ForceEnglishOutput -NonInteractive
```
Comment (daschuer) [inline .github/workflows/build.yml:194]: let' use something more descriptive like "pkg-github-mixxx" or whatever describes the URL the best.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -192,11 +180,32 @@ jobs:
...
+        echo "VCPKG_BINARY_SOURCES=clear;nuget,mixxx,readwrite;" >> "${GITHUB_ENV}"
+
+    - name: Get Git commit of vcpkg submodule
+      if: runner.os != 'Linux'
+      shell: bash
+      run: echo VCPKG_COMMIT=$(git ls-tree HEAD vcpkg | awk '{print $3}') >> "${GITHUB_ENV}"
```
Comment (daschuer) [inline .github/workflows/build.yml:200]: We should not reuse the VCPKG prefix.

File: tools/debian_buildenv.sh [hunk 6]
```diff
@@ -80,6 +80,7 @@ case "$1" in
             libusb-1.0-0-dev \
             libwavpack-dev \
             markdown \
+            ninja-build \
```
Comment (daschuer) [inline tools/debian_buildenv.sh:83]: I don't think we should use Ninja on Linux. Let's stick with the default main. 
At least this is not related to this PR. We may discuss it in a separate PR.

File: .gitmodules (hunk 1 above)
Comment (daschuer) [reply .gitmodules:3]: > If you don't need the submodule then simply don't clone it. It is only needed for macOS and Windows.

That's why we have the scripts in the tools directory. Let's stick to this approach and a vcpkg tree outside the mixxx tree.

File: CMakeLists.txt [hunk 7]
```diff
@@ -26,13 +26,56 @@ if(POLICY CMP0071)
...
+if(VCPKG)
+  message(STATUS "Using dependencies from vcpkg")
+
   set(X_VCPKG_APPLOCAL_DEPS_INSTALL TRUE)
 
+  if(NOT DEFINED ENV{VCPKG_BINARY_SOURCES})
+    set(ENV{VCPKG_BINARY_SOURCES} "clear;nuget,mixxx,read;default;")
+  endif()
+
+  set(ENV{VCPKG_DISABLE_COMPILER_TRACKING} "1")
```
Comment (daschuer) [inline CMakeLists.txt:44]:
```suggestion
  set(ENV{VCPKG_DISABLE_COMPILER_TRACKING} "true")
```

File: CMakeLists.txt [hunk 8]
```diff
@@ -26,13 +26,56 @@ if(POLICY CMP0071)
...
+      if(NOT GIT_FOUND)
+        find_package(Git)
+      endif()
+
+      if(GIT_FOUND)
+        message(STATUS "Initializing vcpkg Git submodule")
+        execute_process(
+          COMMAND ${GIT_EXECUTABLE} submodule init
+          WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
+      )
```
Comment (daschuer) [inline CMakeLists.txt:68]: indent

File: CMakeLists.txt [hunk 9]
```diff
@@ -26,13 +26,56 @@ if(POLICY CMP0071)
...
+        execute_process(
+          COMMAND ${GIT_EXECUTABLE} submodule update
+          WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
+      )
```
Comment (daschuer) [inline CMakeLists.txt:72]: indent

File: CMakeLists.txt [hunk 10]
```diff
@@ -26,13 +26,56 @@ if(POLICY CMP0071)
...
+    set(VCPKG_ROOT "${CMAKE_SOURCE_DIR}/vcpkg")
+
+    if(NOT EXISTS "${VCPKG_ROOT}/bootstrap-vcpkg.sh")
+
+      # Make sure we have actually attempted to find Git
+      if(NOT GIT_FOUND)
+        find_package(Git)
+      endif()
+
+      if(GIT_FOUND)
```
Comment (daschuer) [inline CMakeLists.txt:63]: please consider to inverse this condition to have the error message closer to GIT_FOUND

File: CMakeLists.txt [hunk 11]
```diff
@@ -26,13 +26,56 @@ if(POLICY CMP0071)
...
+        execute_process(
+          COMMAND ${GIT_EXECUTABLE} submodule init
+          WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
+      )
+        execute_process(
+          COMMAND ${GIT_EXECUTABLE} submodule update
+          WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
+      )
+      else()
+        message(FATAL_ERROR "Unable to initialize vcpkg Git submodule because CMake was unable to find a git installation")
```
Comment (daschuer) [inline CMakeLists.txt:74]:
```suggestion
        message(FATAL_ERROR "Unable to initialize vcpkg Git submodule because CMake was unable to find a git executable")
```

File: CMakeLists.txt [hunk 12]
```diff
@@ -53,6 +96,8 @@ if(DEFINED VCPKG_ROOT)
   if(NOT DEFINED CMAKE_TOOLCHAIN_FILE)
     set(CMAKE_TOOLCHAIN_FILE "${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake" CACHE STRING "")
   endif()
+else()
+  message(STATUS "Using dependencies from system without vcpkg")
```
Comment (daschuer) [inline CMakeLists.txt:100]: Remove? This is not relevant for Linux.

**Review by daschuer — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 7 above)
Comment (daschuer) [reply CMakeLists.txt:44]: Do we want that for the GitHub Workflow case? I don't think so

File: CMakeLists.txt (hunk 12 above)
Comment (Be-ing) [reply CMakeLists.txt:100]: It is relevant for Linux. vcpkg can be used on Linux and this is helpful to show whether it is being used or not.

File: CMakeLists.txt (hunk 7 above)
Comment (Be-ing) [reply CMakeLists.txt:44]: This is required for the packages built on GitHub Actions to be usable locally.

File: tools/debian_buildenv.sh (hunk 6 above)
Comment (Be-ing) [reply tools/debian_buildenv.sh:83]: Before this was handled by the buildenv scripts. But now those are not used. It is simpler to just add `-G Ninja` for all conditions.

File: .github/workflows/build.yml (hunk 5 above)
Comment (Be-ing) [reply .github/workflows/build.yml:200]: Why not?

File: .github/workflows/build.yml [hunk 13]
```diff
@@ -192,11 +180,32 @@ jobs:
...
+      uses: nuget/setup-nuget@v1
+      if: runner.os != 'Linux'
+      with:
+        nuget-version: 5.10.0
+
+    - name: "Authenticate to GitHub Packages"
+      if: runner.os != 'Linux' && github.repository_owner == 'mixxxdj'
+      shell: bash
+      run: |
+        nuget sources add -Name mixxx -Source https://nuget.pkg.github.com/mixxxdj/index.json -Username mixxxdj -Password ${{ secrets.GITHUB_TOKEN }} -StorePasswordInClearText -ForceEnglishOutput -NonInteractive
```
[the remark this answers is not in the document]
Comment (Be-ing) [reply .github/workflows/build.yml:193]: I think so. https://vcpkg.readthedocs.io/en/latest/users/binarycaching/#github-packages

File: .gitmodules (hunk 1 above)
Comment (Be-ing) [reply .gitmodules:3]: I don't think "I don't like submodules" is a good enough reason to not do this. It is how vcpkg is designed to be used.

File: .github/workflows/build.yml (hunk 5 above)
Comment (Be-ing) [reply .github/workflows/build.yml:200]: Please suggest another name then.

File: .gitmodules (hunk 1 above)
Comment (Be-ing) [reply .gitmodules:3]: As vcpkg adds new features they are likely to be unavailable with our legacy setup. This is already the case with pinning versions.

File: CMakeLists.txt [hunk 14]
```diff
@@ -26,13 +26,56 @@ if(POLICY CMP0071)
...
+if(VCPKG)
+  message(STATUS "Using dependencies from vcpkg")
+
   set(X_VCPKG_APPLOCAL_DEPS_INSTALL TRUE)
 
+  if(NOT DEFINED ENV{VCPKG_BINARY_SOURCES})
+    set(ENV{VCPKG_BINARY_SOURCES} "clear;nuget,mixxx-github-packages,read;default;")
+  endif()
+
+  set(ENV{VCPKG_DISABLE_COMPILER_TRACKING} "true")
```
Comment (daschuer) [inline CMakeLists.txt:44]: This must be only active for local builds. We want to rebuild the packages on the CI if the compiler version changes. 
Imagine a compile bug is fixed. We don't want to have that in our release builds. For local builds the user can decide to trust the CI or rebuild vcpkg with there own compiler version.

Comment (Be-ing) [reply CMakeLists.txt:44]: > For local builds the user can decide to trust the CI or rebuild vcpkg with there own compiler version.

This is not correct. If VCPKG_DISABLE_COMPILER_TRACKING is not used on CI, local builds will never have cache hits from CI packages.

Comment (daschuer) [reply CMakeLists.txt:44]: Ah OK, makes sense.

File: CMakeLists.txt (hunk 12 above)
Comment (daschuer) [reply CMakeLists.txt:100]: OK.

File: tools/debian_buildenv.sh (hunk 6 above)
Comment (daschuer) [reply tools/debian_buildenv.sh:83]: We can make it conditionally. But again, this change is unrelated to this PR.

File: .gitmodules (hunk 1 above)
Comment (daschuer) [reply .gitmodules:3]: I am am not opposing to the solution at a whole. I am against using the git submodul feature for this. 

We can easily replace this with a normal git call. Submodules are quite unhandy, especially if the submodule is only required for some targets. The submodule feature does not work well with git worktrees.
In this case it becomes worse, because of the size of the "installed*" folder which is ignored by git anyway. 
But it is exactly that folder which contains our requirements.  

In addition  directory structure like this will be much more convenient. 

- workspace
-- mixxx_main
-- mixxx_fetaure1
-- mixxx_fetaure2
-- vcpkg_2.4 

If we use a script from the tools directory or dedicated git commands from the CMakeList.txt for downloading or building the vcpkg folders we have much more chance to control the behaviour and inform the contributors.   
This can contain the commit hash in the same way the submodul feature does in addition to the full download, or whatever fits the best.  

We have discussed using submodules for libraries in our lib folder like here: 
https://github.com/mixxxdj/mixxx/pull/2753#issuecomment-655324002

Comment (Be-ing) [reply .gitmodules:3]: This seems like an elaborate overcomplication that would be worse than using a submodule. This is a different use case from using a submodule for vendored libraries.

> The submodule feature does not work well with git worktrees.

What do you mean? I just tested creating a git worktree from the Tenacity repo. The vcpkg submodule was in the worktree but was not automatically initialized by git, which I think is fine. My use cases for git worktree with Mixxx are working on controller mappings and skins in separate branches from C++, so cloning the vcpkg submodule into those worktrees wouldn't be relevant. So what is the issue?

Comment (daschuer) [reply .gitmodules:3]: I use work trees for feature branches. If I where on MacOs a central vckpg folder would be convenient for this.

File: .github/workflows/build.yml (hunk 2 above)
Comment (daschuer) [reply .github/workflows/build.yml:137]: At least no topic of this PR.

File: .gitmodules (hunk 1 above)
Comment (Be-ing) [reply .gitmodules:3]: You could still do that by explicitly setting the VCPKG_ROOT CMake variable.

File: .github/workflows/build.yml (hunk 5 above)
Comment (daschuer) [reply .github/workflows/build.yml:214]: > Why not?

I have started to craw the VCPKG manuals to find out what this is for. We should release others from this work. 

Let's drop the idea of using submodules ..

Comment (Be-ing) [reply .github/workflows/build.yml:214]: How about MIXXX_VCPKG_SUBMODULE_COMMIT?

Comment (daschuer) [reply .github/workflows/build.yml:214]: Please read the commit form a file per target, and then check out the package with a normal git command. 
This maintains all flexibility and should fullfill your needs as well without introducing the issues known form git submodules.

Comment (daschuer) [reply .github/workflows/build.yml:214]: this would be MIXXX_VCPKG_COMMIT

Comment (Be-ing) [reply .github/workflows/build.yml:214]: That would be reimplementing Git submodule but hackier.

File: .gitmodules (hunk 1 above)
Comment (Be-ing) [reply .gitmodules:3]: Setting VCPKG_ROOT is hardly different from what you would have to do to compile different Git worktrees on macOS and Windows already. Either you'd need to redownload and extract a >1 GB ZIP file for each worktree with the buildenv script or you'd need to manually set CMAKE_TOOLCHAIN_FILE to one central location.

File: .github/workflows/build.yml (hunk 5 above)
Comment (daschuer) [reply .github/workflows/build.yml:214]: How do target the issue if we need different hashes for different targets? 

Using a repository with submodules is not trivial, I really like to avoid the related headaches for all of us. Git submodules is oversized to just download a folder.

Comment (Be-ing) [reply .github/workflows/build.yml:214]: > How do target the issue if we need different hashes for different targets?

We don't do that. I consider that a problematic aspect of our current setup. If one OS needs something different from another, then that needs to be specified either in vcpkg.json or using an overlay. I do not think maintaining separate branches of the vcpkg repo for different OSes is a good idea.

**Comment by Be-ing** [comment]
Superseded by #4432
